### PR TITLE
refactor(iota-types): extract Multiaddr into the iota-multiaddr crate

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

`iota-types` depended on `iota-network-stack` only to re-export its `multiaddr` module. That put the three-level `iota-tls → iota-http → iota-network-stack` tail underneath `iota-types`, which 67 workspace crates wait for.

`multiaddr.rs` has no workspace dependencies of its own, so it becomes a leaf crate. Four commits:

**1. Extract the crate.** Add `iota-multiaddr` holding the module unchanged, re-export it from `iota-network-stack` as `multiaddr` so that crate's consumers are unaffected, and depend on it from `iota-types` instead of `iota-network-stack`. `parse_dns`, `parse_ip4` and `parse_ip6` become `pub`, since `iota-network-stack`'s client uses them across the new boundary. `document_multiaddr_limitation_for_unix_protocol` moves in from `iota-network-stack` — it covers `Multiaddr` rather than the server, and relies on a `#[cfg(test)]` constructor that cannot be reached across a crate boundary.

**2. Move the wasm stand-in into the crate.** `iota-types` was carrying `wasm_multiaddr.rs`, a substitute for a type it doesn't own, and picking between the two with a `cfg` on the re-export. The crate now owns that: `native` and `wasm` modules selected in `lib.rs`, with the native-only dependencies (`anemo`, `eyre`, `idna`, `multiaddr`, `tracing`) gated behind `cfg(not(target_arch = "wasm32"))` — the stand-in needs only `serde`.

**3. Stop re-exporting from `iota-types`.** With the switch inside the crate, `pub use iota_multiaddr as multiaddr;` was pure convenience that hid which crates use the type. Removed; the crates that use `Multiaddr` now import `iota_multiaddr::Multiaddr` and declare the dependency: `iota-config`, `iota-core`, `iota`, `iota-tool`, `iota-swarm-config`, `iota-network`, `iota-node`, `iota-genesis-builder`. `iota-types` keeps the dependency, since `Multiaddr` appears in its own public API via `NetworkMetadata` and `EpochStartValidatorInfoV1`. Crates reaching the type through `iota-network-stack`'s re-export are unchanged — that one lives in the crate that owns the networking stack.

**4. Docs.** Reword the crate summary and drop two intra-doc links that cannot resolve (`wasm` is `cfg`'d out of the native build; `super::Multiaddr` resolved to the stand-in itself on wasm).

The crate's public API differs by target: the parsing helpers and `to_anemo_address` are native-only. `lib.rs` and the `wasm` module docs both say so.

## Build graph effect

Measured with `cargo metadata`, counting workspace crates only and excluding `cfg(msim)` / `wasm32`-gated edges:

| | before | after |
| --- | --- | --- |
| longest compilation chain | 16 | **15** |
| `iota-types` | 3 | **2** |
| `iota-execution` | 6 | 5 |
| `iota-genesis-common` | 7 | 6 |
| `iota-core` | 10 | 9 |
| `iota-node` | 12 | 11 |
| `test-cluster` | 14 | 13 |

`iota-types` now sits at level 2, floored only by `iota-macros` and `iota-protocol-config`. All 67 of its transitive dependents shift down with it.

For context on why this edge specifically: removing every edge in the graph one at a time, this is the only remaining single edge whose removal shortens the critical path and that is not inherent to the design. The other two candidates a greedy search finds are `iota-execution → iota-adapter-latest` (the execution layer dispatching to its adapter) and `test-cluster → iota-swarm` (test-cluster's whole purpose).

Commits 3 and 4 have no build-graph effect — all eight crates already sat above `iota-multiaddr` at level 0. They are about honest dependencies and docs.

## Links to any relevant issues

Part of #12234.

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

- `cargo check --workspace --all-targets`: zero errors, zero warnings — worth the full sweep given `iota-types`' reach
- `cargo test -p iota-multiaddr`: 15 passed, 0 failed (the module's own tests plus the relocated one)
- `iota-multiaddr`, `iota-types` and `iota-config` all build for `wasm32-unknown-unknown`; `iota-config` is the consumer that actually exercises the stand-in
- `cargo doc --no-deps` warning-free on both the native and `wasm32-unknown-unknown` targets
- `cargo machete`, `scripts/cargo_sort/run_consolidate.sh`, `cargo clippy --all-targets` and `cargo +nightly fmt --check` all clean

#### Breaking Changes Rollout

`iota_types::multiaddr` no longer exists — import `iota_multiaddr::Multiaddr` instead. `iota_network_stack::{multiaddr, Multiaddr}` resolves as before. Three previously `pub(crate)` parse helpers are now `pub` on `iota-multiaddr`. All workspace callers are updated in this PR and none of the crates involved is published.

### CODE REVIEW

File: crates/iota-types/src/lib.rs [hunk 1]
```diff
@@ -10,7 +10,7 @@
 )]
 
 #[cfg(not(target_arch = "wasm32"))]
-pub use iota_network_stack::multiaddr;
+pub use iota_multiaddr as multiaddr;
```
Comment (thibault-martinez) [inline crates/iota-types/src/lib.rs]: Why even re-export it? Couldn't we maybe drop more iota-types deps if we don't import multiaddr through it?

Comment (Thoralf-M) [reply crates/iota-types/src/lib.rs]: The re-export is the wasm switch rather than a convenience: `iota_types::multiaddr` resolves to `iota_multiaddr` on native and to the `wasm_multiaddr.rs` stand-in on wasm32. `iota-config` builds for wasm and imports `iota_types::multiaddr::Multiaddr` in `local_ip_utils.rs` and `persisted_config.rs` (both ungated), so importing `iota_multiaddr` directly there would pull in `anemo` and break its wasm build.

It also wouldn't let `iota-types` drop any dependency, because it uses `Multiaddr` in its own public API — `NetworkMetadata` in `committee.rs`, `EpochStartValidatorInfoV1`, and `iota_system_state_summary`. `pub mod committee` isn't wasm-gated, so a `Multiaddr` type is needed on every target either way.

The only thing tying `iota-multiaddr` to `anemo` is `Multiaddr::to_anemo_address`, so in principle moving that out would make the crate wasm-safe and let the stand-in and the `cfg` split go. I don't think it's worth it:

- no build-graph effect — `anemo` is an external crate, so it never counted toward workspace levels
- no compile-volume effect — both consumers of `iota-multiaddr` already depend on `anemo` directly, and `iota-types` needs it independently for `anemo::{PeerId, types::{PeerAffinity, PeerInfo}}`
- it wouldn't remove the statement-level `#[cfg(not(target_arch = "wasm32"))]` gates around the `to_anemo_address` checks in `iota_system_state_inner_v1`, since those gate a node-only check by design

That leaves ~70 lines deleted and the removal of one asymmetry: the real impl validates on deserialize (`s.parse()`), the stand-in doesn't. The wire format is the same in both (a serde string), so it's a validation difference, not a compatibility one — and per the comment in `iota_system_state_inner_v1`, wasm is inspecting already-on-chain state that the Move verifier gated at validator creation. Happy to revisit if that gap is a concern, but it costs ~19 call sites to fix for little gain.

---
_Generated by [Claude Code](https://claude.ai/code)_

Comment (thibault-martinez) [reply crates/iota-types/src/lib.rs]: I feel like the wasm stub should be in the new crate though

Comment (Thoralf-M) [reply crates/iota-types/src/lib.rs]: Agreed, that's better — done in 1de5d15b. The crate now owns the switch:

```rust
// iota-multiaddr/src/lib.rs
#[cfg(not(target_arch = "wasm32"))]
mod native;
#[cfg(target_arch = "wasm32")]
mod wasm;

#[cfg(not(target_arch = "wasm32"))]
pub use native::*;
#[cfg(target_arch = "wasm32")]
pub use wasm::*;
```

`wasm_multiaddr.rs` moved in as the `wasm` module, the native-only dependencies (`anemo`, `eyre`, `idna`, `multiaddr`, `tracing`) are gated behind `cfg(not(target_arch = "wasm32"))`, and the stand-in needs only `serde`. `iota-types` now just has `pub use iota_multiaddr as multiaddr;` with no `cfg` split, and its dependency moved out of the non-wasm section into `[dependencies]`.

One thing worth flagging for review: the crate's public API still differs by target, since the parsing helpers and `to_anemo_address` are native-only. `lib.rs` and the `wasm` module docs both say so, and I moved the note about the missing deserialize-side validation into the `wasm` module where the two implementations now sit next to each other.

Verified: `cargo check --workspace --all-targets` clean with no warnings; `iota-multiaddr`, `iota-types` and `iota-config` all build for `wasm32-unknown-unknown`; `cargo machete`, cargo-sort, clippy, rustfmt clean; 15 tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

Comment (thibault-martinez) [reply crates/iota-types/src/lib.rs]: Now that this is done, does it change anything to my initial question? I'm not fully sure why we would go through iota-types

File: crates/iota-multiaddr/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Mysten Labs, Inc.
+// Modifications Copyright (c) 2026 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! A `Multiaddr` newtype over the `multiaddr` crate.
```
Comment (thibault-martinez) [inline crates/iota-multiaddr/src/lib.rs]: this is a bit weird

File: crates/iota-types/src/lib.rs (hunk 1 above)
Comment (Thoralf-M) [reply crates/iota-types/src/lib.rs]: Yes — it invalidates my earlier answer. The re-export was justified as the wasm switch, and moving the stand-in into the crate removed that reason, leaving pure convenience. Dropped in 6697af01.

`iota-types` no longer re-exports `multiaddr`; the crates that use `Multiaddr` now import `iota_multiaddr::Multiaddr` and declare the dependency themselves. That turned out to be wider than I first estimated — I'd said 4 sites in 3 crates, but multi-line grouped imports hid the rest, and it's 16 sites across 8 crates: `iota-config`, `iota-core`, `iota`, `iota-tool`, `iota-swarm-config`, `iota-network`, `iota-node`, `iota-genesis-builder`.

`iota-types` keeps the dependency, since `Multiaddr` appears in its own public API via `NetworkMetadata` and `EpochStartValidatorInfoV1`. Crates that reach the type through `iota-network-stack`'s re-export (`client.rs`, `server.rs`, and starfish's `tonic_network.rs`) are untouched — that re-export is within the crate that owns the networking stack, so it isn't the same situation.

To be clear about the payoff: no build-graph change. All eight crates already sat above `iota-multiaddr` at level 0, so nothing shifts; this is about each crate declaring what it uses rather than borrowing it through `iota-types`.

Verified: `cargo check --workspace --all-targets` clean, no warnings; `iota-multiaddr`, `iota-types` and `iota-config` all build for `wasm32-unknown-unknown`; `cargo machete`, cargo-sort (stable on re-run), rustfmt clean; 15 tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/iota-multiaddr/src/lib.rs (hunk 2 above)
Comment (Thoralf-M) [reply crates/iota-multiaddr/src/lib.rs]: What about `Multiaddr wrapper for IOTA`?

Comment (Thoralf-M) [reply crates/iota-multiaddr/src/lib.rs]: Taken, thanks — the summary line is now `//! `Multiaddr` wrapper for IOTA.` in 34c2dec6. The old wording restated the crate name twice without saying anything useful.

While checking that, `cargo doc` turned up two intra-doc links I'd added that can't resolve, so they're plain text now:

- `[`wasm`]` in `lib.rs` — the module is `cfg`'d out of the native build, so rustdoc can't see it
- `[`Multiaddr`]: super::Multiaddr` in `wasm.rs` — on wasm the crate root re-exports `wasm::*`, so that resolved to the stand-in itself rather than the native type

`cargo doc --no-deps` is now warning-free for both the native and `wasm32-unknown-unknown` targets.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by thibault-martinez — APPROVED** [review]

**Review by DaughterOfMars — APPROVED** [review]

**Review by valeriyr — APPROVED** [review]

**Review by muXxer — APPROVED** [review]
